sdks/python/apache_beam/io/filesystemio.py

```diff
@@ -0,0 +1,492 @@
+    return self.position
+
+  def _check_open(self):
+    if self.closed:
+      raise IOError('Buffer is closed.')
+
+  def seekable(self):
+    return True
+
+  def readable(self):
+    return True
+
+  def writable(self):
+    return False
+
+
+# TODO: Consider using cStringIO instead of buffers and data_lists when reading
+# and writing.
+class BufferedWriter(object):
+  """A class for writing files from stateless services.
+
+  Provides a Python File object-like interface.
+  """
+
+  class PipeStream(object):
+    """A class that presents a pipe connection as a readable stream."""
+
+    def __init__(self, recv_pipe):
+      self.conn = recv_pipe
+      self.closed = False
+      self.position = 0
+      self.remaining = ''
+
+    def read(self, size):
+      """Read data from the wrapped pipe connection.
+
+      Args:
+        size: Number of bytes to read. Actual number of bytes read is always
+              equal to size unless EOF is reached.
+
+      Returns:
+        data read as str.
+      """
+      data_list = []
+      bytes_read = 0
+      while bytes_read < size:
+        bytes_from_remaining = min(size - bytes_read, len(self.remaining))
+        data_list.append(self.remaining[0:bytes_from_remaining])
+        self.remaining = self.remaining[bytes_from_remaining:]
+        self.position += bytes_from_remaining
+        bytes_read += bytes_from_remaining
+        if not self.remaining:
+          try:
+            self.remaining = self.conn.recv_bytes()
+          except EOFError:
+            break
+      return ''.join(data_list)
+
+    def tell(self):
+      """Tell the file's current offset.
+
+      Returns:
+        current offset in reading this file.
+
+      Raises:
+        IOError: When this stream is closed.
+      """
+      self._check_open()
+      return self.position
+
+    def seek(self, offset, whence=os.SEEK_SET):
+      # The apitools.base.py.transfer.Upload class insists on seeking to the end
+      # of a stream to do a check before completing an upload, so we must have
+      # this no-op method here in that case.
+      if whence == os.SEEK_END and offset == 0:
+        return
+      elif whence == os.SEEK_SET and offset == self.position:
+        return
+      raise NotImplementedError
+
+    def _check_open(self):
+      if self.closed:
+        raise IOError('Stream is closed.')
+
+  def __init__(self,
+               uploader,
+               path,
+               mode='w',
+               mime_type='application/octet-stream'):
+    self.uploader = uploader
+    self.path = path
+    self.mode = mode
+    self.closed = False
+    self.position = 0
+
+    # A small buffer to avoid CPU-heavy per-write pipe calls.
+    self.write_buffer = bytearray()
+    self.write_buffer_size = 128 * 1024
+
+    # Set up communication with uploader.
+    parent_conn, child_conn = multiprocessing.Pipe()
+    self.child_conn = child_conn
+    self.conn = parent_conn
+
+    self.uploader.start(child_conn)
```

Comment:
<!--new_thread; commit:970afe0523d6d5673489a70e801be6553e0ffc33; resolved:0-->
Can we wrap this in a `PipeStream` here (and modify the `Uploader` API accordingly)?

Reply:
`Uploader` API has been changed so `PipeStream` is now an implementation detail of `GcsUploader`.